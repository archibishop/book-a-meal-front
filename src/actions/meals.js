import { MEAL_LIST, MENU_LIST_FAILED} from './types'

export const getMeals = (id, authToken) => dispatch => {
    let payload = {
        method: 'GET',
        headers: {
            'x-access-token': authToken
        }
    }
    
    fetch(`http://127.0.0.1:5000/bookmealapi/v1.0/meals/` + id , payload)
        .then(response => response.json())
        .then(data => 
            dispatch(
            {
                type: MEAL_LIST,
                payload: data.meals
            })
        )
        .catch(error => dispatch(
            {
                type: MENU_LIST_FAILED,
                payload: error
            }))

}


