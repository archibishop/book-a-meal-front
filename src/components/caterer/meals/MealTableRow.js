import React, { Component} from 'react';


export class MealTr extends Component{

    componentWillMount() {
    }
    
    handleEdit = (e) => {
        e.preventDefault();
        this.props.toggleEdit(this.props.meal);
    }

    handleDelete = (e) => {
        e.preventDefault();
        this.props.toggleDelete(this.props.meal.id);
    }
    render(){
        return(
            <tr>
                {/* <td>{ this.props.meal.id }</td> */}
                <td>{ this.props.meal.meal_name }</td>
                <td>{ this.props.meal.meal_type }</td>
                <td>{ this.props.meal.price }</td>
                <td><a id="myBtn" href="./EditMeal.html" className="button-edit" onClick={this.handleEdit}>Edit</a></td>
                <td><button className="button" onClick={this.handleDelete} >Delete</button></td>
            </tr>
        );
    }
}

export default MealTr;
