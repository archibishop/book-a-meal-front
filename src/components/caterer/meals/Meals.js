import React, { Component } from 'react';
import MealTr from './MealTableRow';


class Meals extends Component{

    componentWillMount(){
    }

    renderMeals = () => this.props.meals.map((meal, index)=>(
        <MealTr
            toggleEdit={this.props.toggleEditButton}
            toggleDelete={this.props.toggleDeleteButton}
            key={index}
            meal={meal}
        />
    ));
    render(){
        return(
            <div>
                <table >
                    <tbody>
                    <tr>
                        {/* <th>Id</th> */}
                        <th>Name</th>
                        <th>Type</th>
                        <th>Price</th>
                        <th>Edit</th>
                        <th>Delete</th>
                    </tr>
                    {this.renderMeals()}
                    </tbody>
                </table>
            </div>
        );
    }
}

export default Meals;
