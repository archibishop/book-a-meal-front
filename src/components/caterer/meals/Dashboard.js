import React, { Component } from 'react';
import Navb from '../../navs/Navbar3';
import Meals from './Meals';
import '../../../css/Styles.css';
import Modal from '../../modal/Modal';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { getMeals } from '../../../actions/meals';
import { addMeal } from '../../../actions/addMeal';
import { updateMeal } from '../../../actions/updateMeal';
import { deleteMeal } from '../../../actions/deleteMeal';
import Notifications, { notify } from 'react-notify-toast';

class Dashboard extends Component{

    constructor(props){
        super(props);
        this.state = {
            isOpen: false,
            isOpen1: false,
            isOpen2: false,
            editMeal:{
                meal_name: '',
                price:''
            }
        }
    }
    
    componentWillMount(){
        this.props.getMeals(localStorage.getItem("user_id"),localStorage.getItem("x-access-token"));
    }

    componentWillReceiveProps(data){
        if (JSON.stringify(this.props) !== JSON.stringify(data)) 
        {
            this.getMealList();
        }
    }

    getMealList = () => {
        this.props.getMeals(localStorage.getItem("user_id"),localStorage.getItem("x-access-token"));
    }


    toggleModal = (id) => {
        this.setState({
            isOpen: !this.state.isOpen,
            id: id
        });
    }

    toggleModal1 = (meal) => {
        this.setState({
            isOpen1: !this.state.isOpen1,
            editMeal: meal }, () => {
        });
    }

    toggleModal2 = () => {
        this.setState({
            isOpen2: !this.state.isOpen2,
        });
    }

    handleAdd = (e) => {
        e.preventDefault();
        this.toggleModal2();

    }

    handleAddMeal = (e) => {
        e.preventDefault();
        let mealData = {
            meal_name: e.target.elements.firstname.value,
            price: parseInt(e.target.elements.lastname.value),
            meal_type: e.target.elements.country.value,
            admin_id: parseInt(localStorage.getItem('user_id'))
        }
        this.props.addMeal(JSON.stringify(mealData))
        this.toggleModal2();
    }

    handleEditMeal = (e) => {
        e.preventDefault();
        let mealData = {
            meal_name: e.target.elements.firstname.value,
            price: parseInt(e.target.elements.lastname.value),
            meal_type: e.target.elements.country.value,
            admin_id: parseInt(localStorage.getItem('user_id'))
        }
        this.props.updateMeal(this.state.editMeal.id, JSON.stringify(mealData))
        this.toggleModal1(mealData)
    }

    handleDeleteMeal = (e) => {
        e.preventDefault();
        this.props.deleteMeal(parseInt(this.state.id))
        this.toggleModal(this.state.id)
    }

    render(){
        return(
            <div className="dashboard">
                <Notifications />
                <Navb />
                <div className="container">
                    <h1>MEALS</h1>
                    <a className="button-add" onClick={this.handleAdd}>Add Meal</a>
                    <br />
                </div>
                <Meals meals={this.props.mealList} toggleEditButton={this.toggleModal1.bind(this)} 
                toggleDeleteButton={this.toggleModal.bind(this)} />
                <Modal show={this.state.isOpen}
                    onClose={this.toggleModal}
                    orderList={this.props.orderList}>
                    Are you sure you want delete the Meal.
                    <br />
                    <br />
                    <button id="btn-delete" className="button" onClick={this.handleDeleteMeal}>
                        Delete
                    </button>
                </Modal>
                <Modal show={this.state.isOpen2}
                    onClose={this.toggleModal2}
                    orderList={this.props.orderList}>
                    <h1>Add Meal</h1>

                    <div >
                        <form id="form-add" onSubmit={this.handleAddMeal}>
                            <label htmlFor="fname">Meal</label>
                            <input type="text" id="fname" name="firstname" placeholder="Your name.."  />

                            <label htmlFor="lname">Price</label>
                            <input type="number" id="lname" name="lastname" placeholder="Your last name.." />

                            <label htmlFor="country">Type</label>
                            <select id="country" name="country">
                                <option value="breakfast">Breakfast</option>
                                <option value="lunch">Lunch</option>
                                <option value="supper">Supper</option>
                            </select>

                            <input type="submit" value="Submit" />
                        </form>
                    </div>
                </Modal>
                <Modal show={this.state.isOpen1}
                    onClose={this.toggleModal1}
                    orderList={this.props.orderList}>
                    <h1>Edit Meal</h1>
                    <div >
                        <form id="form-edit" onSubmit={this.handleEditMeal}>
                            <label htmlFor="fname">Meal</label>
                            <input type="text" id="fname" name="firstname" placeholder={this.state.editMeal.meal_name} />

                            <label htmlFor="lname">Price</label>
                            <input type="text" id="lname" name="lastname" placeholder={this.state.editMeal.price} />

                            <label htmlFor="country">Type</label>
                            <select id="country" name="country">
                                <option value="breakfast">Breakfast</option>
                                <option value="lunch">Lunch</option>
                                <option value="supper">Supper</option>
                            </select>

                            <input type="submit" value="Submit"/>
                        </form>
                    </div>
                </Modal>
            </div>
        );
    }
}

Dashboard.propTypes = {
    getMeals: PropTypes.func.isRequired,
    addMeal: PropTypes.func.isRequired,
    updateMeal: PropTypes.func.isRequired,
    deleteMeal: PropTypes.func.isRequired
}

const mapStateToProps = state => ({
    mealList: state.meals.meals,
    addMealInfo: state.addMeal.message,
    editMealInfo: state.updateMeal.message,
    deleteMealInfo: state.deleteMeal.message
})

export default connect(mapStateToProps, {getMeals, addMeal, updateMeal, deleteMeal})(Dashboard);
