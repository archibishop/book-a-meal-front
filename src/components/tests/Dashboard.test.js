import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Dashboard } from '../caterer/meals/Dashboard';
import { MemoryRouter } from 'react-router-dom';
import { debug } from 'util';

const mockStore = configureStore();
const initialState = {
    selectReducer: {
        selectedAvatar: 0,
    },
    getOrders: (x) => {
    },
    orders: [],
    user: [
        {
            message: 'HipHIP',
            error: null,
        }
    ],
};
const store = mockStore(initialState);


describe('<Dashboard />', () => {
    describe('render()', () => {
        test('renders the component', () => {
            const getMealssMock = jest.fn();
            const meals = [];
            const wrapper = shallow(<Dashboard meals={meals} getMeals={getMealssMock} />);

            expect(toJson(wrapper)).toMatchSnapshot();
        });
    });
});

describe('<Dashboard />', () => {
    describe('render()', () => {
        test('test delete button works', () => {
            const getMealssMock = jest.fn();
            const deleteMealMock = jest.fn();
            const meals = [{
                id: 3,
                meal_name: "flash",
                price: 6000,
                meal_type: "Crush"
            }];
            const wrapper = mount(
                <MemoryRouter>
                    <Dashboard mealList={meals} getMeals={getMealssMock} deleteMeal={deleteMealMock}/>
                </MemoryRouter>
                );
            expect(wrapper.find('Meals').length).toEqual(1)

            wrapper.find('MealTr').prop('toggleEdit')({ target: {} })
            wrapper.find('MealTr').prop('toggleDelete')({ target: {} })
            // const component = wrapper.dive();

            let isOpen = true
            wrapper.setProps({ isOpen })
            expect(wrapper.find('Modal').length).toEqual(3)
            expect(wrapper.find('#btn-delete').length).toEqual(1)
            wrapper.find('#btn-delete').simulate('click');
            expect(deleteMealMock.mock.calls.length).toEqual(1);

            
        });
    });
});


describe('<Dashboard />', () => {
    describe('render()', () => {
        test('expects Modals rendered', () => {
            const getMealssMock = jest.fn();
            const addMealMock = jest.fn();
            const meals = [{
                id: 3,
                meal_name: "flash",
                price: 6000,
                meal_type: "Crush"
            }];
            const wrapper = mount(
                <MemoryRouter>
                    <Dashboard mealList={meals} getMeals={getMealssMock} addMeal={addMealMock} />
                </MemoryRouter>
            );
            expect(wrapper.find('Meals').length).toEqual(1)

            wrapper.find('MealTr').prop('toggleEdit')({ target: {} })
            wrapper.find('MealTr').prop('toggleDelete')({ target: {} })

            let isOpen1 = true
            wrapper.setProps({ isOpen1 })
            expect(wrapper.find('Modal').length).toEqual(3)
        });
    });
});

describe('<Dashboard />', () => {
    describe('render()', () => {
        test('form submissions work', () => {
            const getMealssMock = jest.fn();
            const updateMealMock = jest.fn();
            const addMealMock = jest.fn();
            const meals = [];
            const wrapper = shallow(<Dashboard meals={meals} getMeals={getMealssMock} addMeal={addMealMock} updateMeal={updateMealMock}/>);
            // const component = wrapper.dive();


            let isOpen1 = true
            wrapper.setProps({ isOpen1 })

            expect(wrapper.find('Modal').length).toEqual(3)
            expect(wrapper.find('form').length).toEqual(2)
            expect(wrapper.find('#form-edit').length).toEqual(1)
            expect(wrapper.find('#form-add').length).toEqual(1)

            wrapper.find('#form-add').simulate('submit', {
                preventDefault: () => {
                },
                target: {
                    elements : {
                    firstname: {
                        value: "Fast foods"
                    },
                    lastname: {
                        value: "JOHNNY"
                    },
                    country: {
                        value: "123456"
                    }
                }
                }
            });
            expect(addMealMock.mock.calls.length).toEqual(1);

            wrapper.find('#form-edit').simulate('submit', {
                preventDefault: () => {
                },
                target: {
                    elements: {
                        firstname: {
                            value: "Fast foods"
                        },
                        lastname: {
                            value: "JOHNNY"
                        },
                        country: {
                            value: "123456"
                        }
                    }
                }
            });
            expect(updateMealMock.mock.calls.length).toEqual(1);

            wrapper.find('.button-add').simulate('click', {
                preventDefault: () => {
                }
            });
        });
    });
});

