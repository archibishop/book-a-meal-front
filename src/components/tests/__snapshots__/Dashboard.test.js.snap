// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> render() renders the component 1`] = `
<div
  className="dashboard"
>
  <_class
    options={Object {}}
  />
  <Navb />
  <div
    className="container"
  >
    <h1>
      MEALS
    </h1>
    <a
      className="button-add"
      onClick={[Function]}
    >
      Add Meal
    </a>
    <br />
  </div>
  <Meals
    toggleDeleteButton={[Function]}
    toggleEditButton={[Function]}
  />
  <Modal
    onClose={[Function]}
    show={false}
  >
    Are you sure you want delete the Meal.
    <br />
    <br />
    <button
      className="button"
      id="btn-delete"
      onClick={[Function]}
    >
      Delete
    </button>
  </Modal>
  <Modal
    onClose={[Function]}
    show={false}
  >
    <h1>
      Add Meal
    </h1>
    <div>
      <form
        id="form-add"
        onSubmit={[Function]}
      >
        <label
          htmlFor="fname"
        >
          Meal
        </label>
        <input
          id="fname"
          name="firstname"
          placeholder="Meal Name..."
          type="text"
        />
        <label
          htmlFor="lname"
        >
          Price
        </label>
        <input
          id="lname"
          name="lastname"
          placeholder="Price 2000UGX..."
          type="number"
        />
        <label
          htmlFor="country"
        >
          Type
        </label>
        <select
          id="country"
          name="country"
        >
          <option
            value="breakfast"
          >
            Breakfast
          </option>
          <option
            value="lunch"
          >
            Lunch
          </option>
          <option
            value="supper"
          >
            Supper
          </option>
        </select>
        <input
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </Modal>
  <Modal
    onClose={[Function]}
    show={false}
  >
    <h1>
      Edit Meal
    </h1>
    <div>
      <form
        id="form-edit"
        onSubmit={[Function]}
      >
        <label
          htmlFor="fname"
        >
          Meal
        </label>
        <input
          id="fname"
          name="firstname"
          placeholder=""
          type="text"
        />
        <label
          htmlFor="lname"
        >
          Price
        </label>
        <input
          id="lname"
          name="lastname"
          placeholder=""
          type="text"
        />
        <label
          htmlFor="country"
        >
          Type
        </label>
        <select
          id="country"
          name="country"
        >
          <option
            disabled={true}
            selected={true}
            value={true}
          >
             -- select an option -- 
          </option>
          <option
            value="breakfast"
          >
            Breakfast
          </option>
          <option
            value="lunch"
          >
            Lunch
          </option>
          <option
            value="supper"
          >
            Supper
          </option>
        </select>
        <input
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </Modal>
</div>
`;
